eth/protocols/eth/peer.go

```diff
@@ -323,20 +344,159 @@ func (p *Peer) RequestReceipts(hashes []common.Hash, sink chan *Response) (*Requ
+		}
+		header := backend.Chain().GetHeaderByHash(p.requestedReceipts[requestId][lastBlock])
+		logSize, err := p.validateLastBlockReceipt(packet.List, requestId, header)
+		if err != nil {
+			return err
+		}
+
+		// Update the buffered data and trim the packet to exclude the incomplete block.
+		if buffer, ok := p.receiptBuffer[requestId]; ok {
+			// If the buffer is already allocated, it means that the previous response was incomplete
+			// Append the first block receipts
+			buffer.list[len(buffer.list)-1].Append(packet.List[0])
+			buffer.list = append(buffer.list, packet.List[1:]...)
+			buffer.lastLogSize = logSize
+		} else {
+			p.receiptBuffer[requestId] = &partialReceipt{
+				list:        packet.List,
+				lastLogSize: logSize,
+			}
+		}
+		return nil
+	}
+
+	// Request completed
+	if buffer, ok := p.receiptBuffer[requestId]; ok {
+		// If the request is completed, append previously collected receipts
+		// to the packet and remove the buffered receipts.
+		packet.List = append(buffer.list, packet.List...)
+		delete(p.receiptBuffer, requestId)
+	}
+	delete(p.requestedReceipts, requestId)
+
+	return nil
+}
+
+// validateLastBlockReceipt validates receipts and return log size of last block receipt.
+// This function is called only when the `lastBlockincomplete == true`.
+// Note that the last receipt response (which completes receiptLists of a pending block) is not verified here.
+// Those response doesn't need hueristics below since they can be verified by its trie root.
+func (p *Peer) validateLastBlockReceipt(receiptLists []*ReceiptList69, id uint64, header *types.Header) (uint64, error) {
+	lastReceipts := receiptLists[len(receiptLists)-1]
+
+	// If the receipt is in the middle of retreival, use the buffered data.
+	// e.g. [[receipt1], [receipt1, receipt2], incomplete = true]
+	//      [[receipt3, receipt4], incomplete = true] <<--
+	//      [[receipt5], [receipt1], incomplete = false]
+	// This case happens only if len(receiptLists) == 1 && incomplete == true && buffered before.
+	var previousTxs int
+	var previousLog uint64
+	var log uint64
+	if buffer, ok := p.receiptBuffer[id]; ok && len(receiptLists) == 1 {
+		previousTxs = len(buffer.list[len(buffer.list)-1].items)
+		previousLog = buffer.lastLogSize
+	}
+
+	// 1. Verify that the total number of transactions delivered is under the limit.
+	if uint64(previousTxs+len(lastReceipts.items)) > header.GasUsed/21_000 {
+		// should be dropped, don't clear the buffer
+		return 0, fmt.Errorf("total number of tx exceeded limit")
+	}
+
+	// 2. Verify the size of each receipt against the maximum transaction size.
+	for _, rc := range lastReceipts.items {
+		if uint64(len(rc.Logs)) > params.MaxTxGas/params.LogDataGas {
+			return 0, fmt.Errorf("total size of receipt exceeded limit")
+		}
+		log += uint64(len(rc.Logs))
+	}
+	// 3. Verify that the overall downloaded receipt size does not exceed the block gas limit.
+	if previousLog+log > header.GasUsed/params.LogDataGas {
```

Comment:
Is `previousLog+log` correct here in case of requesting receipts across three requests? If I see it correctly, we track the size of the last response in `lastLogSize` but we don't accumulate it?